include/mfmg/dealii/lanczos.templates.hpp

```diff
@@ -280,6 +298,158 @@ Lanczos<OperatorType, VectorType>::details_calc_tridiag_epairs(
+  // of T.  If an eigenvalue of T is not repeated, and it is also
+  // an eigenvalue of T2, then it is considered spurious.
+  //
+  // The C/W algorithm has other nuances which may necessitate more
+  // revisions of the simplified algoithm implemented here.
+
+
+  // The following tolerance may need adjustment, however this doesn't
+  // seem critical (cf. C/W vol. 1). Cullum/Willoughby use 1e-12.
+  const double tol2 = 5.e-12;
+
+  // Identify repeated eigenvalues.
+  // A "marked" eigenvalue here is nonrepeated or first of a set of repeated.
+
+  std::vector<bool> is_repeated(n);
+  std::vector<bool> is_marked(n);
+
+  for (int i = 0; i < n; ++i) {
+    is_repeated[i] = (i>0 && evals[i-1] >= evals[i] - tol2) ||
+                     (i<n-1 && evals[i+1] <= evals[i] + tol2);
+    is_marked[i] = 0==i || evals[i-1] < evals[i] - tol2;
+  } // i
+
+  // Identify spurious eigenvalues.
+
+  std::vector<bool> is_spurious(n);
+  const int n2 = n - 1;
+
+  if (n2 >= 1 && n2 >= num_requested) {
+
+    // Solve an eigenproblem based on deleting the first row and col
+    // of the original tridiag matrix.
+    // (only need evals, not evecs, so a different LAPACK call may be adequate)
+
+    std::vector<double> evals2;
+    std::vector<double> evecs2;
+
+    evals2 = main_diagonal;
+    evals2.erase(evals2.begin());
+    std::vector<double> sub_diagonal_aux = sub_diagonal;
+    sub_diagonal_aux.erase(sub_diagonal_aux.begin());
+    std::vector<double> evecs_aux(n2 * n2);
+
+    const lapack_int info =
+        LAPACKE_dstev(LAPACK_COL_MAJOR, 'V', n2, evals2.data(),
+                    sub_diagonal_aux.data(), evecs_aux.data(), n2);
+    ASSERT(!info, "Call to LAPACKE_dstev failed.");
+
+#ifdef VERBOSE
+    std::cout << "    " << " T2" << " ";
+    for (int i=0; i<num_requested; ++i)
+      std::cout << evals2[i] << " ";
+#endif
+
+    // Loop over original eigenvalues to check for spuriousness..
+    // NOTE: assuming here evals and evals2 are in ascending order.
+
+    int j_start = 0;
+
+    for (int i = 0; i < n; ++i) {
+
+      is_spurious[i] = false;
+
+      // A repeated eigenvalue of T is never spurious.
+
+      if (is_repeated[i])
+        continue;
+
+      // Seek matching T2 eigenvalue. If found, then eval[i] is spurious.
+      // Note the looping is rigged here to avoid an O(n^2) algorithm.
+
+      for (int j = j_start; j < n2; ++j) {
+
+        const bool is_t2j_below = evals2[j] < evals[i] - tol2;
+        if (is_t2j_below) {
+          // For the next i, evals[i] will be >= this one,
+          // so not examining this j for next i will be ok.
+          j_start = j;
+          continue;
+        }
+
+        const bool is_t2j_above = evals2[j] > evals[i] + tol2;
+        if (is_t2j_above)
+          // we have passed up evals[i], so no match.
+          break;
+
+        // in interval surrounding evals[i], thus evals[i] spurious.
+        is_spurious[i] = true;
+        break;
+
+      } // n2
+
+    } // n
```

Comment:
I feel like it might be more readable with STL algorithms:
```C++
auto first = evals.begin();
auto const last = evals.end();
for (int i = 0; i < n; ++i) {
  if (is_repeated[i]) continue;
  first = std::find_if(first, last,
                       [&is_strictly_ordered, &evals, i](double x) {
                         return !is_strictly_ordered(x, evals[i]);
                       });
  if (first != last && !is_strictly_ordered(evals[i], *first)) is_spurious[i] = true;
}
```
@aprokop @Rombur Any opinion on this?

**Disclaimer** I have not actually compiled and tested the pseudo code I wrote :)

Reply:
One thing about the issue of handling the (a-tol, a+tol) interval.  In the CW algorithm some attention is given to making the intervals exactly consistent regarding the multiplicity test and the spuriousness test.  it is possible but I think doubtful that the inclusion or exclusion of the endpoints of the interval would make much difference.  So I think the simplification of the code is probably ok --